### ISSUE TITLE
Fn::ImportValue throws error when used in deploymentBucket

### ISSUE THREAD

**Issue by azjones** [serverless/serverless#5001]
# This is a (Bug Report)
Fn::ImportValue throws error when used in deploymentBucket
## Description
Fn::ImportValue (!ImportValue) ('Fn::ImportValue') when used in
provider:
&nbsp;&nbsp;deploymentBucket:
&nbsp;&nbsp;&nbsp;&nbsp;name:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fn::ImportValue: DeploymentBucket

Where `DeploymentBucket` is a valid export.
## Expectation
I would expect `Fn::ImportValue` to work here as it works in other parts, such as:
provider:
&nbsp;&nbsp;vpc:
&nbsp;&nbsp;&nbsp;&nbsp;securityGroupIds:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;- Fn::ImportValue: 'SecurityGroupId'

## Limitations of workarounds
The current work around is to use
`${cf:<stackName>:<export>}`
The limitation to that is that I must know at all times the name of the `<stackName>` which should be considered dynamic. Deploying to dev, the `<stackName>` may be dev-bucket, where deploying to stage the `<stackName>` may be stage-bucket.

## Additional Data
* ***Operating System***: macOS High Sierra
```
  Serverless Error ---------------------------------------

  Bucket name must start with a letter or number. [object Object]

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Forums:        forum.serverless.com
     Chat:          gitter.im/serverless/serverless

  Your Environment Information -----------------------------
     OS:                     darwin
     Node Version:           8.11.1
     Serverless Version:     1.26.1
```

**Comment by azjones** [comment]
This seems to be an issue specifically with `deploymentBucket`

If you attempt to pass something other than a string to `name:` you get the above Serverless Error.

Also, if you attempt to pass an inline pseudo parameter (while using serverless-pseudo-parameters plugin) like `deploys-#{AWS::AccountId}-${opt:region}` - you get a Serverless Error stating that a bucket name must not contain upper case letters.

**Comment by naingyelin** [comment]
I can confirm this is still occurring for version 1.32.0. I am not able to use the cross stack export of a bucket name as my deploymentBucketName with Fn::ImportValue.
I believe that the validation of the bucket name is occurring before cross stack export reference is resolved.

**Comment by joshuanapoli** [comment]
The problem was previously fixed but rejected in #3189.

It is frustrating that we cannot predict which configuration properties will work with config variables and CF exports.

**Comment by ruthsteen** [comment]
I am seeing the issue with using Fn::Join. We are supposed to use the same bucket for each account for all lambda deployments. Is there no fix or workaround for this?

**Comment by sdomagala** [comment]
As a fix I came up with this helper:
```
// serverless-helpers.js
module.exports.getAccountId = async () => {
    const { Account } = await sts.getCallerIdentity().promise();
    return Account;
};
```

and then in code:
```
// serverless.yml
variable: ${file(./serverless-helpers.js)):getAccountId}
```

and it works just fine.

Docs:
https://serverless.com/framework/docs/providers/aws/guide/variables#reference-variables-in-javascript-files

Although I feel serverless should have this kind of helper on its own as it's simple enough and could save a lot of trouble.

**Comment by ot-alex-vance** [comment]
+1 definitely still a problem

**Comment by mbluemer** [comment]
This is something that's still clearly advertised in the documentation as working: https://www.serverless.com/framework/docs/providers/aws/guide/variables#reference-cloudformation-outputs 😞

**Comment by circulon** [comment]
This is still broken / an issue in 2.4.0 ;(
Building stack dependencies on a comm bucket is just not possible in this manner, which is frustrating.

**Comment by ndt75** [comment]
Definitely still an issue in 01/21. I had to go with the JS config files route to make it work.

**Comment by pkit** [comment]
> This is something that's still clearly advertised in the documentation as working: https://www.serverless.com/framework/docs/providers/aws/guide/variables#reference-cloudformation-outputs disappointed

Yep, this one is just unbelievable as the documentation clearly states that it should work, but it doesn't!

**Comment by pgrzesik** [comment]
@pkit Could you clarify where the docs are wrong in this case? `deploymentBucket` is a tricky one, as it has to be resolvable outside of CloudFormation, that's why it cannot be provided as CF intrinsic functions

**Comment by pkit** [comment]
@pgrzesik I could not successfully use `Fn::ImportValue` anywhere I've tried.
For example in `provider.env`

**Comment by pgrzesik** [comment]
Hello @pkit Could you please share a small reproducible example that is failing for you? For `provider.environment` it should work just fine.

**Comment by pkit** [comment]
Works:
```
provider:
  environment:
    DATA_API_SECRET_ARN: ${cf:stack-name-${self:provider.stage}.SecretArn}
```
Doesn't work:
```
provider:
  environment:
    DATA_API_SECRET_ARN:
        'Fn::ImportValue': 'stack-name-${self:provider.stage}-SecretArn'
```
Complains that validation is failed, not a string.

The definition in the stack is:
```
Outputs:
  SecretArn:
    Description: 'Secret Arn'
    Value: !Ref Secret
    Export:
      Name: !Sub '${AWS::StackName}-SecretArn'
```

**Comment by pgrzesik** [comment]
Could you please share the detailed output of the `package` or `deploy` command for failing case? I've tried to reproduce but it works without issues in the exact same example as yours. Are you sure you're using latest version of the Framework @pkit ?

**Comment by pkit** [comment]
@pgrzesik it looks like the error comes from plugins, as they don't have access to the same variables if they don't run in the `deploy:deploy` lifecycle stage (see here https://github.com/serverless/serverless/pull/3911).

**Comment by pgrzesik** [comment]
Hello @pkit  - do you have an example of a specific plugin and command at which it happens? I'd like to see if I can reproduce on my side

**Comment by pmuen** [comment]
Still open :(
